tls/src/main/java/org/bouncycastle/tls/crypto/impl/jcajce/JceTlsKyberDomain.java

```diff
@@ -0,0 +1,89 @@
+import org.bouncycastle.pqc.crypto.crystals.kyber.KyberKEMGenerator;
+import org.bouncycastle.pqc.crypto.crystals.kyber.KyberKeyGenerationParameters;
+import org.bouncycastle.pqc.crypto.crystals.kyber.KyberKeyPairGenerator;
+import org.bouncycastle.pqc.crypto.crystals.kyber.KyberParameters;
+import org.bouncycastle.pqc.crypto.crystals.kyber.KyberPrivateKeyParameters;
+import org.bouncycastle.pqc.crypto.crystals.kyber.KyberPublicKeyParameters;
+import org.bouncycastle.tls.NamedGroup;
+import org.bouncycastle.tls.crypto.TlsAgreement;
+import org.bouncycastle.tls.crypto.TlsKEMConfig;
+import org.bouncycastle.tls.crypto.TlsKEMDomain;
+
+public class JceTlsKyberDomain implements TlsKEMDomain
+{
+    public static KyberParameters getKyberParameters(TlsKEMConfig kemConfig)
+    {
+        switch (kemConfig.getKEMNamedGroup())
+        {
+        case NamedGroup.kyber512:
+            return KyberParameters.kyber512;
+        case NamedGroup.kyber768:
+            return KyberParameters.kyber768;
+        case NamedGroup.kyber1024:
+            return KyberParameters.kyber1024;
+        default:
+            return null;
+        }
+    }
+
+    protected final JcaTlsCrypto crypto;
+    protected final TlsKEMConfig kemConfig;
+    protected final KyberParameters kyberParameters;
+
+    public TlsKEMConfig getTlsKEMConfig()
+    {
+        return kemConfig;
+    }
+
+    public JceTlsKyberDomain(JcaTlsCrypto crypto, TlsKEMConfig kemConfig)
+    {
+        this.crypto = crypto;
+        this.kemConfig = kemConfig;
+        this.kyberParameters = getKyberParameters(kemConfig);
+    }
+
+    public TlsAgreement createKEM()
+    {
+        return new JceTlsKyber(this);
+    }
+
+    public KyberPublicKeyParameters decodePublicKey(byte[] encoding)
+    {
+        return new KyberPublicKeyParameters(kyberParameters, encoding);
+    }
+
+    public byte[] encodePublicKey(KyberPublicKeyParameters kyberPublicKeyParameters)
+    {
+        return kyberPublicKeyParameters.getEncoded();
+    }
+
+    public AsymmetricCipherKeyPair generateKeyPair()
+    {
+        KyberKeyPairGenerator keyPairGenerator = new KyberKeyPairGenerator();
+        keyPairGenerator.init(new KyberKeyGenerationParameters(crypto.getSecureRandom(), kyberParameters));
+        return keyPairGenerator.generateKeyPair();
+    }
+
+    public JceTlsSecret adoptLocalSecret(byte[] secret)
+    {
+        return crypto.adoptLocalSecret(secret);
+    }
+
+    public SecretWithEncapsulation enCap(KyberPublicKeyParameters peerPublicKey)
+    {
+        KyberKEMGenerator kemGen = new KyberKEMGenerator(crypto.getSecureRandom());
+        return kemGen.generateEncapsulated(peerPublicKey);
+    }
+
+    public byte[] deCap(KyberPrivateKeyParameters kyberPrivateKeyParameters, byte[] cipherText)
+    {
+        KyberKEMExtractor kemExtract = new KyberKEMExtractor(kyberPrivateKeyParameters);
+        byte[] secret = kemExtract.extractSecret(cipherText);
+        return secret;
```

Comment:
As mentioned in https://github.com/bcgit/bc-java/issues/1578#issuecomment-1929400882, we can add a wrapper to the new Kyber encapsulation and decapsulation methods in order to support the NIST round 3 Kyber implementation.

```suggestion
       byte[] outputSharedSecret = new byte[32];
       byte[] kr = new byte[64];
       System.arraycopy(secret, 0, kr, 0, 32);
       hash_h(kr, cipherText, 32);
       kdf(outputSharedSecret, kr);
       return outputSharedSecret;
```